src/remote/blobs.go

```diff
@@ -0,0 +1,335 @@
+			// Continue and send all of these, it is not necessarily fatal (although it
+			// will probably blow up later, but easier not to handle that error here)
+			for _, b := range blobs {
+				chOut <- b
+			}
+			return
+		}
+		for _, d := range resp.MissingBlobDigests {
+			chOut <- m[d.Hash]
+		}
+	}
+
+	// Buffer them up a bit, the request supports checking multiple at once which is
+	// likely more efficient than doing them all one at a time.
+	blobs := make([]*blob, 0, buffer)
+	for b := range chIn {
+		// The actual hash might or might not be set.
+		if len(b.Digest.Hash) == 0 {
+			h, err := c.state.PathHasher.Hash(b.File, false, true)
+			if err != nil {
+				return err
+			}
+			b.Digest.Hash = hex.EncodeToString(h)
+		}
+		blobs = append(blobs, b)
+		if len(blobs) == buffer {
+			filter(blobs)
+			blobs = blobs[:0]
+		}
+	}
+	if len(blobs) != 0 {
+		filter(blobs)
+	}
+	close(chOut)
+	return g.Wait()
+}
+
+// reallyUploadBlobs actually does the upload of the individual blobs, after they have
+// been filtered through FindMissingBlobs.
+func (c *Client) reallyUploadBlobs(ch <-chan *blob) error {
+	reqs := []*pb.BatchUpdateBlobsRequest_Request{}
+	var totalSize int64
+	for b := range ch {
+		if b.Digest.SizeBytes > c.maxBlobBatchSize {
+			// This blob individually exceeds the size, have to use this
+			// ByteStream malarkey instead.
+			if err := c.storeByteStream(b); err != nil {
+				return err
+			}
+		} else if b.Digest.SizeBytes+totalSize > c.maxBlobBatchSize {
+			// We have exceeded the total but this blob on its own is OK.
+			// Send what we have so far then deal with this one.
+			if err := c.sendBlobs(reqs); err != nil {
+				return err
+			}
+			reqs = []*pb.BatchUpdateBlobsRequest_Request{}
+			totalSize = 0
+		}
+		// This file is small enough to be read & stored as part of the request.
+		// Similarly the data might or might not be available.
+		if b.Data == nil {
+			data, err := ioutil.ReadFile(b.File)
+			if err != nil {
+				return err
+			}
+			b.Data = data
+		}
+		reqs = append(reqs, &pb.BatchUpdateBlobsRequest_Request{
+			Digest: b.Digest,
+			Data:   b.Data,
+		})
+	}
+	if len(reqs) > 0 {
+		return c.sendBlobs(reqs)
+	}
+	return nil
+}
+
+// sendBlobs dispatches a set of blobs to the remote CAS server.
+func (c *Client) sendBlobs(reqs []*pb.BatchUpdateBlobsRequest_Request) error {
+	ctx, cancel := context.WithTimeout(context.Background(), reqTimeout)
+	defer cancel()
+	resp, err := c.storageClient.BatchUpdateBlobs(ctx, &pb.BatchUpdateBlobsRequest{
+		InstanceName: c.instance,
+		Requests:     reqs,
+	})
+	if err != nil {
+		return err
+	}
+	// TODO(peterebden): this is not really great handling - we should really use Details
+	//                   instead of Message (since this ends up being user-facing) and
+	//                   shouldn't just take the first one. This will do for now though.
+	for _, r := range resp.Responses {
+		if r.Status.Code != 0 {
```

Comment:
Is there an enumerated set of response codes? I'm guessing `0` means Success?

Reply:
I think it's the same as the gRPC ones. Inexplicably it is not actually an enum though - a comment in the file says it's a member of `google.rpc.Code` but it's actually just an `int32`. Not sure what's going on but I'll use one of those.